### ISSUE TITLE
Type instantiation is excessively deep not reported with a complex type involving a conditional type and another aliased conditional type

### ISSUE THREAD

**Issue by Andarist** [microsoft/TypeScript#50017]
# Bug Report

### 🔎 Search Terms

NonNullable, conditional type, excessively deep

### 🕗 Version & Regression Information

- This changed between versions [4.2](https://www.typescriptlang.org/play?ts=4.2.3#code/C4TwDgpgBAysBOBLAdgcwNIRAZwDwHkA+KAXigFEAPBAQwGNhcBrLAewDMp8AaKbBFKkIBuAFChIUAGo0ANgFcIeIqS4BtFiA5cAumPHhoASXa4Awq2QATRMESXeAFQAWEZL3KzsEYmQvXbe2QoCGo3K2woBEUoAH4oFzcoAC4KLwgxCWNsfAAjACsIBlxHXwSQsOtIgCUi1ngrXH4kNF4aZBBieOjoVPY5b0zDKCNsAEF4eBoQErLHCuBwyImpmfbOuKj4GL6BjNFRFEX4frpoAA0aYBpaunqrKABvUSgoRCtU5sExAF8DrKgAFkxucAPrVchmACq1RgRnwADlVABWfQAhFKRZWCwKAC2yGUvFu8ng2EQADcIAAFK7OBZLKDrNQ6XwvKAAenZUWciEidFcdCYb0iiFxYHq12QwCirFYbOJpIp1NpagARLI3KhgM5VTp6VUgSDwZCYXDEWzXvFkBBKfALSlpHJFHh7a9Xs83Z63Wp0G9ggARK40KnwVhgZSEHSpEy4V1e16jFbTAg+lncOPx9Dp+Pxznc3lQBGWBHyWSyGi5DXCt5iiXtYAZr1F5AlssVjUp9As-URKCXa63e6N+PxPMY-gQbGsPEEqAACngRRJZMpAEooK5F3wIBBcZFgKwoLloKLxfBJQ2c1fC5jJzj5PiXdfr3ntQXm63y5WTyLa+f68O14fqWX4dvgqaENmz45nmuSsNqUDaNqrDeIyW6nnWUqAVeagAHT4QqK7KtqvDgV2OjYV6xChIsBpfGglGevEAAGAAkjzjli96PrG0FAcWIHthAnZpoxMFcnBCFIc4KHQDQ6F-heYnxnhBFLoqlI0iR6jkcpbqED8zF6ak1q2mJJk2hAdrPiIjZ-PGtkGJIgbXCGYYRqoVB0AoVjCZo2j4FAABkfACK0UD+Zw-Y3HUDS2XmtghJM9SRAA7rYdJvpE5LyYgAGvsMLnBqG4YEGUkVcMFoUtKgYhAA) and [4.3](https://www.typescriptlang.org/play?ts=4.3.5#code/C4TwDgpgBAysBOBLAdgcwNIRAZwDwHkA+KAXigFEAPBAQwGNhcBrLAewDMp8AaKbBFKkIBuAFChIUAGo0ANgFcIeIqS4BtFiA5cAumPHhoASXa4Awq2QATRMESXeAFQAWEZL3KzsEYmQvXbe2QoCGo3K2woBEUoAH4oFzcoAC4KLwgxCWNsfAAjACsIBlxHXwSQsOtIgCUi1ngrXH4kNF4aZBBieOjoVPY5b0zDKCNsAEF4eBoQErLHCuBwyImpmfbOuKj4GL6BjNFRFEX4frpoAA0aYBpaunqrKABvUSgoRCtU5sExAF8DrKgAFkxucAPrVchmACq1RgRnwADlVABWfQAhFKRZWCwKAC2yGUvFu8ng2EQADcIAAFK7OBZLKDrNQ6XwvKAAenZUWciEidFcdCYb0iiFxYHq12QwCirFYbOJpIp1NpagARLI3KhgM5VTp6VUgSDwZCYXDEWzXvFkBBKfALSlpHJFHh7a9Xs83Z63Wp0G9ggARK40KnwVhgZSEHSpEy4V1e16jFbTAg+lncOPx9Dp+Pxznc3lQBGWBHyWSyGi5DXCt5iiXtYAZr1F5AlssVjUp9As-URKCXa63e6N+PxPMY-gQbGsPEEqAACngRRJZMpAEooK5F3wIBBcZFgKwoLloKLxfBJQ2c1fC5jJzj5PiXdfr3ntQXm63y5WTyLa+f68O14fqWX4dvgqaENmz45nmuSsNqUDaNqrDeIyW6nnWUqAVeagAHT4QqK7KtqvDgV2OjYV6xChIsBpfGglGevEAAGAAkjzjli96PrG0FAcWIHthAnZpoxMFcnBCFIc4KHQDQ6F-heYnxnhBFLoqlI0iR6jkcpbqED8zF6ak1q2mJJk2hAdrPiIjZ-PGtkGJIgbXCGYYRqoVB0AoVjCZo2j4FAABkfACK0UD+Zw-Y3HUDS2XmtghJM9SRAA7rYdJvpE5LyYgAGvsMLnBqG4YEGUkVcMFoUtKgYhAA) and I've bisected it to [this PR](https://github.com/microsoft/TypeScript/pull/43527)

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.8.0-dev.20220723#code/C4TwDgpgBAysBOBLAdgcwNIRAZwDwHkA+KAXigFEAPBAQwGNhcBrLAewDMp8AaKbBFKkIBuAFChIUAGo0ANgFcIeIqS4BtFiA5cAumPHhoASXa4Awq2QATRMESXeAFQAWEZL3KzsEYmQvXbe2QoCGo3K2woBEUoAH4oFzcoAC4KLwgxCWNsfAAjACsIBlxHXwSQsOtIgCUi1ngrXH4kNF4aZBBieOjoVPY5b0zDKCNsAEF4eBoQErLHCuBwyImpmfbOuKj4GL6BjNFRFEX4frpoAA0aYBpaunqrKABvUSgoRCtU5sExAF8DrKgAFkxucAPrVchmACq1RgRnwADlVABWfQAhFKRZWCwKAC2yGUvFu8ng2EQADcIAAFK7OBZLKDrNQ6XwvKAAenZUWciEidFcdCYb0iiFxYHq12QwCirFYbOJpIp1NpagARLI3KhgM5VTp6VUgSDwZCYXDEWzXvFkBBKfALSlpHJFHh7a9Xs83Z63Wp0G9ggARK40KnwVhgZSEHSpEy4V1e16jFbTAg+lncOPx9Dp+Pxznc3lQBGWBHyWSyGi5DXCt5iiXtYAZr1F5AlssVjUp9As-URKCXa63e6N+PxPMY-gQbGsPEEqAACngRRJZMpAEooK5F3wIBBcZFgKwoLloKLxfBJQ2c1fC5jJzj5PiXdfr3ntQXm63y5WTyLa+f68O14fqWX4dvgqaENmz45nmuSsNqUDaNqrDeIyW6nnWUqAVeagAHT4QqK7KtqvDgV2OjYV6xChIsBpfGglGevEAAGAAkjzjli96PrG0FAcWIHthAnZpoxMFcnBCFIc4KHQDQ6F-heYnxnhBFLoqlI0iR6jkcpbqED8zF6ak1q2mJJk2hAdrPiIjZ-PGtkGJIgbXCGYYRqoVB0AoVjCZo2j4FAABkfACK0UD+Zw-Y3HUDS2XmtghJM9SRAA7rYdJvpE5LyYgAGvsMLnBqG4YEGUkVcMFoUtKgYhAA)

### 💻 Code

```ts
type StringKeys<O> = Extract<keyof O, string>;
type Values<O> = O[keyof O];

type If<Condition, Then, Else> = Condition extends true ? Then : Else;
type IsObject<T> = T extends Record<string, any> ? true : false;
type IsArray<T> = T extends Array<any> ? true : false;

interface XataRecord {
  id: string;
}

type MAX_RECURSION = 5;

type NestedColumns<O, RecursivePath extends any[]> =
  // this check is important too
  RecursivePath["length"] extends MAX_RECURSION
    ? never
    : Values<
        {
          [K in DataProps<O>]: If<
            IsArray<O[K]>,
            K,
            // this NonNullable is important
            NonNullable<O[K]> extends XataRecord
              ? // NestedColumns (recursive) here seems to be important
                NestedColumns<
                  // this NonNullable is important
                  NonNullable<O[K]>,
                  // both of those are important
                  [...RecursivePath, O[K]]
                > extends string
                ? `${NestedColumns<
                    NonNullable<O[K]>,
                    // both of those are important
                    [...RecursivePath, O[K]]
                  >}`
                : never
              : never
          >;
        }
      >;

type DataProps<O> = Exclude<keyof O & string, keyof XataRecord>;
// it errors with this variant
// type DataProps<O> = keyof O & string;
```

### 🙁 Actual behavior

It doesn't error on the template literal type.

### 🙂 Expected behavior

It **should** error on the template literal type.

I believe that it should error because:
1. it was an error in 4.2
2. a bigger repro of this started to error in nightly versions since the `NonNullable` change (from a conditional type to `T & {}`). The slimmed down version that I've created doesn't error in the nightly build though. Original repro by @SferaDev can be found here: [4.7 with no error](https://www.typescriptlang.org/play?ts=4.7.4#code/C4TwDgpgBAysBOBLAdgcwNIRAZwDwHkA+KAXigFEAPBAQwGNhcBrLAewDMp8AaKbBFKkIBuALAAoUJCgA5AK4BbAEYR4mHAWJkqtBszaceUZIpXwREqdABqNADZyIeIqS4BtOEjTrnhALpi4pbg0ACqyIisyAAqrACSyMCq2BAMkci40VpQABTRUBDUEMgAJthQNMggUAD8uZQAXFDRAJSkxJXVTcgQAG6qbYVJpeU5jVAo7KpQAEptJB1VtbNQ3X2qgcHScey4AMJRJYjA6bzRABbFvOR2KdkHpcfpBUUjUAiOyxfFqxS3EJtJCEoHFsPglAArVKMLKufJDYplWapVjwEq4fheVC8TrEOofaBNdj2FKBKwg7AAQXg8BoIEy2XhryR1Np9NxywJv2J-0B5PIR2ANCUdggABEaEKCC9hkiAEI0FISoXZADeEigUDc6AmyCgLBAHC4fia+G1fhliPKAA1JTQZii0RrNctVRMSk1MYIoABfZ2apoyKLyOx2YWigjm4gIt62oUOuiokr+13uz0CNC+qAAH2MclDOagclKEHYKAgyfELt+ZvQAQkPsBAHoAFQtjUt5FgeBOYrAcqVd19xBl8PQHuJtFQdjwVgKd6XKAlO1KRUQAB0HabEhQSXgxLo0DiJWHo5F0HVVagrfbV6gnfCiAAjp9ECUoEbgOdEOUJ0nN3eLbbleb7pligR+kE4i7qoB7QAqSp2lAl6am4BpgYIJoVFUEESBIN4dl2PYpIkA5QJA8DYD+STvn+U4znOC7QMuQqrikAH3sBMH7vQ0BxvajrvjGSLHqeiBjshzoEYBUAALIQEKLE0B+nBfj+UB0SUHGakBzqUHaTQoS60nVvesimNMX6ShM5QoHQPZrkJ-TwNUJwKNAX7joJNlFmALEVtpLq6XeUDOVRUTdBZ8CBJqjZ4VeJk6ciAjrGRPYzk4lzvomYDVJ+i50HINJ9hp3kMfOnlLiua6BcFmoOSUOQtE0AAKs4KD+EC4A6NAlFEdj0jAECigwFYtYgdBMLganYLwbgAOQtvNfiEMQuYmKGFhQZqiVmS1sGogoZFgDQ8AnPYvn+SpTFQIVxWJKVk5aVwerYHIdCHtgZE9AA7h+kLQjZ-qdj2wBFT076VO+lV3T2D2sADDA+cgrDAFAChyEKNG1VAAACJ20vOSnNAVrAOAoyADqUTGIPAoX2I45RWWj5w0P07ysFAKiXZKAXA3joPg+UlLGBAf2abdURCig3qVWGST8PTDhONOqIVAWMNk4olPXTDRVw2jmm1cBmpyH5vM5CmBPnXYoTm0krWnTbuACscY7KjQBAdYwM28AaRr8QmSarYQzrNVAbVzp13UQL1-WDcN0JjRNU2+1qi3LSHgI7W2-P7fuh1kSeoonFEusFfrJUSygN1KWxG7A-zuNfrOP3lDsN0Sz9irqw11TFwpFYfnTI7YQUNJq6zYCQBDUA-d+orkQd8AdZmlUD6XyDG86G8QE1rXtdHvSsG+W2QVs0AADIoFNQdTsJNp2nfJTZPg3u4IHgm8PNDXzYW81m38vNYgAAySSCVc4yQdClfoaVSzESypLXK5doCwyrmVdqtdqrsWBibUqvV94R0PikGOcdkADVwENEaNFxqTRjk9OaGcVprTzJtbO15IFBQjsvI6FRyJO3EnYHmSQUG3Urg9I2L0+DvU+t9MW-0oRIx-PzQW8BZ5Qxumg+GiM0bqRRmjDGWM+YyXxqdGgRMkLfEluTHWGjPK0yVozBc1lWbs2AJzbmgDeZaSbqonWItfqPSTJLRINAZZr0XPLJwaNegMxVuwNW9ghGaxseUfKqDxGG0EtvK8XikiWxCtbQRdt-KOzOoIl2gp3Z2i9scehSY-YGCgJ-J6Icw4HyjiQnqfVyEJ2ocnOhz9GFLWYWfZsTYI5yBFBNJooQTh2GOK5YE7ioCoAUnw4szwfrHHODdFI1CJKJlSddQcCNFHAAvrARODAxwHBsQQXgCYipUX6C1SUuyH7YRAG4C0ZAflaCkhMykCye4JLpp0cis4KKgGdLmDOgKoDAvEuUMF7pcj6N1HwOglwFA0BaLCqA803zzQRZHaFiAVbpLEfdNGop+h2AJR7MlvgEUyGiUPbsrByWUtUpE1gP1ph0uGtgAlbL+AVjudrZwjzUjPMQK895W18ITJalMhZdBZnzMWe8ZZnM1lo0qi04JSg5CIDsGjcqfCFmKyNPspO2UtYUxFUCaQVD7W0KmvgS0sYn5fygOob1SI3U3PPJKimmh-muCNVOEBCKABiatY7YusdrXgArVnrOOOUHo4r3yxOVthd8EIT56mOBUKmuo9wpDSFEZ0j4oixASFW6E6RcAplsMrPARlqzakxeoNwJhlCqD8FhMVNEO2OEpMAMNGQjDoFAc031T0Youh9KHK8SrxBNhVWqmZUA5lmu1eSFZ+qbr5s+EaZSRztYVDRspVA8qfgnS-JcidEAp0zoeRHQN5Rg2sVFJ+og2QWo-sJUtZ0dQ32+GvECkFKLElLE5dCkAzpWqgaJSUElV46hekzNuxFcHVYj3fDkDFNdsDYogLi-FV40OfP9oYCD7gWoWnw2S1QJweWaMyVAIVDLaPfs+QAAwACSqkmNMdAPp1xiYk3TawPohNMZ1PRpp+AUyQbiXgFM1YgzIBDGGc8kY6zRmZOUZ+GIMzYi+eu0yplIOgb-bcx1GQ9MGbHMZ5hOm7OpmsE0N9H6XO4Dc-mQzEZawrV4NYYgkEfOmTWM5bz1ZTTmm87Zl0CXVCodFs5cZRD5W8yaLECox83y3VpoVMMdMTwzxPMgOgFLZpz2gAs72ndZWUWeDVl9LroCyUpNaAA+jMcgexQgzBgHEfAMhXAAFY8ttQKw7fdWrQA6ukCe9ZylNll22V+TRLnygqDsPy7jNLePrH4-hqgdAHAnnW9AGuO3kCZfgJcsdErDtfqeZ1hV+3PmdEjWQH7LyIBvK-AtUUaAvzLVA-1obI2xsTamzIJjPREsCZ2G2kKmpQTgnOZobg3moO4G7aZXtNcmVQt8FhLHSWXSglZHSYLwZQseYi6tInOPTLoF4PhjulUkMceqOpQcp02RpugPwLlj05VRHXFAAA6sxKI800ZvWnqiNGZzAbi7pOUEACl6e492MbhnYIdEs-02zozHPCBc7i7p1noZ2dRlA9Gysjv7NXIOaGoLIWXe26jHNdcoeQePvB+cXgEWVqgdw6gM3dm6g6lzKJ1UUmZOqic37+5AewtdQ5yHsPHXQeR+j+aFainE-xZy1l7ncWmhp4z2JzwggfBW-c0HkzLCM6Kb5xMi4ovQz8vKHIbANA1nXR1wwJmrM709mMKjXjN9nVe81OgM36W7NbTs2uh3pl0d1+rJutjSBYnLYPQstbx69XrOn2jIX5SVZ7fOKwTGi74zeSiE4S5VOuXQe0EoFuzkBPEoSs3byIF4Fb28CwDwA91WgWzP0KxW0PWv11QzQNUXHPWgEvRTQplvT4QfX6D1GfV2VI3ZXfHv0cQgBo3wwTTpgACIElWB1xVx4BWCaAAAvBg5rUqMGNRZCVWVgQyLmU6EQ1cTgpockH0LMSCckD7EoALadILOdLARza5f9CAQDVaUgBFDuK9FzHyZSHNGiSFLlYXXgOTWQQtOSZ0ANYTWTZAKYOmGQaTRw5wuSKvEKOoGbT5P-MAFlbnOoMnOzCnPUGQLCWSdQ33ADf3Z3fPTzVaavHw+I13Ezd3JdJMavTUOoeg7uNETvcLc0XgBQpQz9PPNIyLOSBdeA7I34PI06EoQogvYo2QCgsouI63QPIokzXgWSJI+vGvYsJgFGH6ZAYQGDITdjM6aoMTWSGQvRJfO1ENReMuckMTKTJTQY2LIY5AEY-lPUbdKYqFYXKAMTVw5GJfZSR-NbT8YEMTfALw5Lf1NQvwu0ZlTQDTcBOLMIl4kAUdVIrvWPT5D3M3XI1EfIpom3HovwKAjQgZKaCooE+3dOEZAYr3FLOsFdasHYgMP49DcDbw74nzX46AjAWAzQAErohIjnDIz-J6MEqABogo6ElousXga+PYjvVkxInQ-DTkpgIjSE38WOEoJsPJaAdyL8Vgd8KiVAZASUIqFWFZNmE+d8E7VgKiCJR7RIVQBU5JEIM3TE+sbnXE34YY0Y8YyY6YtbDYrABY2yZAfNN8ITPkYEZkqE7orqaIR5RkUDJEiMWEOoGYX4EM9aULN06QUIFIOmMgN0BU9yDCNAXgJIcxGoIrWOBQZ+X0MkYEaITM1weM8xCAJM6zA41QdM-dGM7MuKXrKs1QbMsgaM6YMBD3XM6QfM8xRs5oAs1szItEQEE8W7U6VBKIRWMfVQJoP9BE3AZs+AIZdOVMhQdcNEwIRMSmA1TM-AMY1QY8VwCc9gpcmodccs+AY8t8CQIAA), [4.8 nightly with the error](https://www.typescriptlang.org/play?ts=4.8.0-dev.20220723#code/C4TwDgpgBAysBOBLAdgcwNIRAZwDwHkA+KAXigFEAPBAQwGNhcBrLAewDMp8AaKbBFKkIBuALAAoUJCgA5AK4BbAEYR4mHAWJkqtBszaceUZIpXwREqdABqNADZyIeIqS4BtOEjTrnhALpi4pbg0ACqyIisyAAqrACSyMCq2BAMkci40VpQABTRUBDUEMgAJthQNMggUAD8uZQAXFDRAJSkxJXVTcgQAG6qbYVJpeU5jVAo7KpQAEptJB1VtbNQ3X2qgcHScey4AMJRJYjA6bzRABbFvOR2KdkHpcfpBUUjUAiOyxfFqxS3EJtJCEoHFsPglAArVKMLKufJDYplWapVjwEq4fheVC8TrEOofaBNdj2FKBKwg7AAQXg8BoIEy2XhryR1Np9NxywJv2J-0B5PIR2ANCUdggABEaEKCC9hkiAEI0FISoXZADeEigUDc6AmyCgLBAHC4fia+G1fhliPKAA1JTQZii0RrNctVRMSk1MYIoABfZ2apoyKLyOx2YWigjm4gIt62oUOuiokr+13uz0CNC+qAAH2MclDOagclKEHYKAgyfELt+ZvQAQkPsBAHoAFQtjUt5FgeBOYrAcqVd19xBl8PQHuJtFQdjwVgKd6XKAlO1KRUQAB0HabEhQSXgxLo0DiJWHo5F0HVVagrfbV6gnfCiAAjp9ECUoEbgOdEOUJ0nN3eLbbleb7pligR+kE4i7qoB7QAqSp2lAl6am4BpgYIJoVFUEESBIN4dl2PYpIkA5QJA8DYD+STvn+U4znOC7QMuQqrikAH3sBMH7vQ0BxvajrvjGSLHqeiBjshzoEYBUAALIQEKLE0B+nBfj+UB0SUHGakBzqUHaTQoS60nVvesimNMX6ShM5QoHQPZrkJ-TwNUJwKNAX7joJNlFmALEVtpLq6XeUDOVRUTdBZ8CBJqjZ4VeJk6ciAjrGRPYzk4lzvomYDVJ+i50HINJ9hp3kMfOnlLiua6BcFmoOSUOQtE0AAKs4KD+EC4A6NAlFEdj0jAECigwFYtYgdBMLganYLwbgAOQtvNfiEMQuYmKGFhQZqiVmS1sGogoZFgDQ8AnPYvn+SpTFQIVxWJKVk5aVwerYHIdCHtgZE9AA7h+kLQjZ-qdj2wBFT076VO+lV3T2D2sADDA+cgrDAFAChyEKNG1VAAACJ20vOSnNAVrAOAoyADqUTGIPAoX2I45RWWj5w0P07ysFAKiXZKAXA3joPg+UlLGBAf2abdURCig3qVWGST8PTDhONOqIVAWMNk4olPXTDRVw2jmm1cBmpyH5vM5CmBPnXYoTm0krWnTbuACscY7KjQBAdYwM28AaRr8QmSarYQzrNVAbVzp13UQL1-WDcN0JjRNU2+1qi3LSHgI7W2-P7fuh1kSeoonFEusFfrJUSygN1KWxG7A-zuNfrOP3lDsN0Sz9irqw11TFwpFYfnTI7YQUNJq6zYCQBDUA-d+orkQd8AdZmlUD6XyDG86G8QE1rXtdHvSsG+W2QVs0AADIoFNQdTsJNp2nfJTZPg3u4IHgm8PNDXzYW81m38vNYgAAySSCVc4yQdClfoaVSzESypLXK5doCwyrmVdqtdqrsWBibUqvV94R0PikGOcdkADVwENEaNFxqTRjk9OaGcVprTzJtbO15IFBQjsvI6FRyJO3EnYHmSQUG3Urg9I2L0+DvU+t9MW-0oRIx-PzQW8BZ5Qxumg+GiM0bqRRmjDGWM+YyXxqdGgRMkLfEluTHWGjPK0yVozBc1lWbs2AJzbmgDeZaSbqonWItfqPSTJLRINAZZr0XPLJwaNegMxVuwNW9ghGaxseUfKqDxGG0EtvK8XikiWxCtbQRdt-KOzOoIl2gp3Z2i9scehSY-YGCgJ-J6Icw4HyjiQnqfVyEJ2ocnOhz9GFLWYWfZsTYI5yBFBNJooQTh2GOK5YE7ioCoAUnw4szwfrHHODdFI1CJKJlSddQcCNFHAAvrARODAxwHBsQQXgCYipUX6C1SUuyH7YRAG4C0ZAflaCkhMykCye4JLpp0cis4KKgGdLmDOgKoDAvEuUMF7pcj6N1HwOglwFA0BaLCqA803zzQRZHaFiAVbpLEfdNGop+h2AJR7MlvgEUyGiUPbsrByWUtUpE1gP1ph0uGtgAlbL+AVjudrZwjzUjPMQK895W18ITJalMhZdBZnzMWe8ZZnM1lo0qi04JSg5CIDsGjcqfCFmKyNPspO2UtYUxFUCaQVD7W0KmvgS0sYn5fygOob1SI3U3PPJKimmh-muCNVOEBCKABiatY7YusdrXgArVnrOOOUHo4r3yxOVthd8EIT56mOBUKmuo9wpDSFEZ0j4oixASFW6E6RcAplsMrPARlqzakxeoNwJhlCqD8FhMVNEO2OEpMAMNGQjDoFAc031T0Youh9KHK8SrxBNhVWqmZUA5lmu1eSFZ+qbr5s+EaZSRztYVDRspVA8qfgnS-JcidEAp0zoeRHQN5Rg2sVFJ+og2QWo-sJUtZ0dQ32+GvECkFKLElLE5dCkAzpWqgaJSUElV46hekzNuxFcHVYj3fDkDFNdsDYogLi-FV40OfP9oYCD7gWoWnw2S1QJweWaMyVAIVDLaPfs+QAAwACSqkmNMdAPp1xiYk3TawPohNMZ1PRpp+AUyQbiXgFM1YgzIBDGGc8kY6zRmZOUZ+GIMzYi+eu0yplIOgb-bcx1GQ9MGbHMZ5hOm7OpmsE0N9H6XO4Dc-mQzEZawrV4NYYgkEfOmTWM5bz1ZTTmm87Zl0CXVCodFs5cZRD5W8yaLECox83y3VpoVMMdMTwzxPMgOgFLZpz2gAs72ndZWUWeDVl9LroCyUpNaAA+jMcgexQgzBgHEfAMhXAAFY8ttQKw7fdWrQA6ukCe9ZylNll22V+TRLnygqDsPy7jNLePrH4-hqgdAHAnnW9AGuO3kCZfgJcsdErDtfqeZ1hV+3PmdEjWQH7LyIBvK-AtUUaAvzLVA-1obI2xsTamzIJjPREsCZ2G2kKmpQTgnOZobg3moO4G7aZXtNcmVQt8FhLHSWXSglZHSYLwZQseYi6tInOPTLoF4PhjulUkMceqOpQcp02RpugPwLlj05VRHXFAAA6sxKI800ZvWnqiNGZzAbi7pOUEACl6e492MbhnYIdEs-02zozHPCBc7i7p1noZ2dRlA9Gysjv7NXIOaGoLIWXe26jHNdcoeQePvB+cXgEWVqgdw6gM3dm6g6lzKJ1UUmZOqic37+5AewtdQ5yHsPHXQeR+j+aFainE-xZy1l7ncWmhp4z2JzwggfBW-c0HkzLCM6Kb5xMi4ovQz8vKHIbANA1nXR1wwJmrM709mMKjXjN9nVe81OgM36W7NbTs2uh3pl0d1+rJutjSBYnLYPQstbx69XrOn2jIX5SVZ7fOKwTGi74zeSiE4S5VOuXQe0EoFuzkBPEoSs3byIF4Fb28CwDwA91WgWzP0KxW0PWv11QzQNUXHPWgEvRTQplvT4QfX6D1GfV2VI3ZXfHv0cQgBo3wwTTpgACIElWB1xVx4BWCaAAAvBg5rUqMGNRZCVWVgQyLmU6EQ1cTgpockH0LMSCckD7EoALadILOdLARza5f9CAQDVaUgBFDuK9FzHyZSHNGiSFLlYXXgOTWQQtOSZ0ANYTWTZAKYOmGQaTRw5wuSKvEKOoGbT5P-MAFlbnOoMnOzCnPUGQLCWSdQ33ADf3Z3fPTzVaavHw+I13Ezd3JdJMavTUOoeg7uNETvcLc0XgBQpQz9PPNIyLOSBdeA7I34PI06EoQogvYo2QCgsouI63QPIokzXgWSJI+vGvYsJgFGH6ZAYQGDITdjM6aoMTWSGQvRJfO1ENReMuckMTKTJTQY2LIY5AEY-lPUbdKYqFYXKAMTVw5GJfZSR-NbT8YEMTfALw5Lf1NQvwu0ZlTQDTcBOLMIl4kAUdVIrvWPT5D3M3XI1EfIpom3HovwKAjQgZKaCooE+3dOEZAYr3FLOsFdasHYgMP49DcDbw74nzX46AjAWAzQAErohIjnDIz-J6MEqABogo6ElousXga+PYjvVkxInQ-DTkpgIjSE38WOEoJsPJaAdyL8Vgd8KiVAZASUIqFWFZNmE+d8E7VgKiCJR7RIVQBU5JEIM3TE+sbnXE34YY0Y8YyY6YtbDYrABY2yZAfNN8ITPkYEZkqE7orqaIR5RkUDJEiMWEOoGYX4EM9aULN06QUIFIOmMgN0BU9yDCNAXgJIcxGoIrWOBQZ+X0MkYEaITM1weM8xCAJM6zA41QdM-dGM7MuKXrKs1QbMsgaM6YMBD3XM6QfM8xRs5oAs1szItEQEE8W7U6VBKIRWMfVQJoP9BE3AZs+AIZdOVMhQdcNEwIRMSmA1TM-AMY1QY8VwCc9gpcmodccs+AY8t8CQIAA) 
3. If we **inline** the `If<IsArray<O[K]>, ..., ...>` alias then it starts to error in all of the versions of the slimmed down repro: [TS playground](https://www.typescriptlang.org/play?ts=4.7.4&ssl=20&ssc=32&pln=20&pc=39#code/C4TwDgpgBAysBOBLAdgcwNIRAZwDwHkA+KAXigFEAPBAQwGNhcBrLAewDMp8AaKbBFKkIBuAFChIUAGo0ANgFcIeIqS4BtFiA5cAumPHhoASXa4Awq2QATRMESXeAFQAWEZL3KzsEYmQvXbe2QoCGo3K2woBEUoAH4oFzcoAC4KLwgxCWNsfAAjACsIBlxHXwSQsOtIgCUi1ngrXH4kNF4aZBBieOjoVPY5b0zDKCNsAEF4eBoQErLHCuBwyImpmfbOuKj4GL6BjNFRFEX4frpoAA0aYBpaunqrKABvUSgoRCtU5sExAF8DrKgAFkxucAPrVchmACq1RgRnwADlVABWfQAhFKRZWCwKAC2yGUvFu8ng2EQADcIAAFK7OBZLKDrNQ6XwvKAAenZUWciEidFcdCYb0iiFxYHq12QwCirFYbOJpIp1NpagARLI3KhgM5VTp6VUgSDwZCYXDEWzXvFkBBKfALSlpHJFHh7a9Xs83Z63Wp0G9ggARK40KnwVhgZSEHSpUYraYEH0s-URLaKV1ey1QdBp9OpTnc3lQBGWBHyWSyGi5DXCt5iiXtYDZr1F5AlssVjXx9CJ0KLA2Xa63e6Nz3xPMY-gQbGsPEEqAACngRRJZMpAEooK5F3wIBBcZFgKwoLloKLxfBJQ301fC5jJzj5PiXdfr3ntQXm63y5WTyLa+f68OV4fqWX4dvgCaENwgHpnmuSsNqUDaNqrDeIyW6nnWUrQV6agAHT4QqK7KtqvDgV2OjYa8xA9gyXxoJRmwAAYACSPOOWL3o+uAMZ6wFtt+nYslBz7PrB8F0khzgodANDoX+F48d6+G4YRSo0iR6jkYphA-IxDGpNatrYQZNoQPAYjXn86YiP8wyBtcIZhhGqhUHQChWBAzBsJw+BQAAZHwAitFAmjaP2Nx1A0Nl5rYISTPUkQAO62HSb6ROSsmIABr52UGjnhgQZShT5-mBS0qBiEAA). To the best of my knowledge the aliased version and inlined version should not differ in their observable behaviors.

I understand that the repro is not small - I've tried to reduce it further but I couldn't. The problem at hand seems to be quite complex and involves a very specific ingredients from what I can tell.

**Comment by RyanCavanaugh** [comment]
> It should error on the template literal type

Wait, what?

Generally we do lots of effort to avoid issuing this error; it's always bad. Why is issuing the error desirable here?

**Comment by SferaDev** [comment]
> Generally we do lots of effort to avoid issuing this error; it's always bad. Why is issuing the error desirable here?

Yeah, actually my original repro was to try to remove this error for good. And that's when @Andarist found that there was a behavioral difference between both options.

**Comment by Andarist** [comment]
> Generally we do lots of effort to avoid issuing this error; it's always bad. Why is issuing the error desirable here?

I don't mind this not being an error. The only thing that I care about it here is consistency. I've just figured out that it might be preferable (or at least easier) to make this an error rather than to make the error go away for all of those cases.

**Comment by RyanCavanaugh** [comment]
I don't think there's anything to fix here. For any potential analytical situation where there's a problem (infinite instantiation) that is sometimes avoidable through additional smarts, it'll always be possible to write two similar programs on either side of that boundary line. If anyone wants to send a PR to move that boundary in the good direction that's fine, but the existence of the boundary itself can't be considered a defect.

**Comment by Andarist** [comment]
Hm, so is it expected that sometimes using an equivalent type through an alias can move the mentioned boundary? The main problem I've seen here is that using a type alias `If` didn't produce the error when the **equivalent** "inlined" conditional type did. Is this expected? 

You are mentioning that it will always be possible to write two similar programs and land them on either side of that boundary. I agree with that. I'd just like to make a last check if those two programs are considered to be similar because I've considered them to be the same in the same sense that those are the same:
```ts
// program1
const result = 2 + 2

// program2
function add(a, b) { return a + b }
const result = add(2, 2)
```

If extracting parts a more complex type to type aliases (without changing its traits, such as homomorphism, distributivity, etc, in the process) is considered to produce a *different*/similar program then I'm OK with closing this.

**Comment by RyanCavanaugh** [comment]
> Hm, so is it expected that sometimes using an equivalent type through an alias can move the mentioned boundary?

Sure, since type aliases can have their variances measured (thus potentially allowing a shortcut measure of the relation between `F<T>` and `F<S>` via `T -> S`), but inline type expressions cannot.
